### DISCUSSION TITLE
External radio, direct modulation "module"

### DISCUSSION THREAD

**Discussion by ManoDaSilva** [jgromes/RadioLib#646] [Ideas]
It would be interesting to create a "module" that would use the direct transmission mode (for any 2-FSK mode) to feed it into an external analog FM radio (such as what is done for most of high power POCSAG transmitters, or for the MMDVM) :)

**Comment by jgromes** [comment]
I'm not sure what exactly you're referring to. Direct mode in the context of sub-GHz transceivers like SX1278 takes clock and data as digital signals on two pins and transmits that without any packetization. I don't think that's something an analog FM radio can deal with. In addition, direct mode is used for some hacks in RadioLib (to transmit AFSK signals by applying square wave tone to the data input).

Do you mean to output just the tone from AFSK on some pin?

**Reply by ManoDaSilva** [reply]
I got the idea when dissecting the inner workings of your POCSAG implementation:
If you look into [unipager](https://github.com/rwth-afu/UniPager), which is a software that turns a Rpi + Analog radio into a POCSAG transmitter, it uses the audio card to generate a square wave tone directly fed into the FM modulator input to create the required 2-FSK (some adjustment on the audio level is required to get the appropriate deviation). 

That's why I think it could be interesting to see if we could do the same with your library, and create a module that would just use the packetization you implemented. I'll play around with it next week to see if I can send the DIO0(?) data signal straight into an FM radio. 

(AFSK should also be doable I guess)

**Reply by jgromes** [reply]
Actually, AFSK is the only one that can be implemented in this way. If you take a carefull look at the implementation of `transmitDirect` for various modules, you will find out it actually accesses the device registers to enable unmodulated carrier output on some radio frequency - you will find nothing on the DIO pins. So what happens is that this frequency is changed by RadioLib during transmission, and so it's the library, not the transceiver that does the modulation in this case.

AFSK is different, since it actually outputs signal to the DIO2 pin, and the carrier is modulated by the transceiver. It is going to work with everything that implements AFSK (Morse, RTTY, SSTV etc.), but not for Pagers, since those only use 2-FSK.

**Reply by jgromes** [reply]
@ManoDaSilva FYI I added support for external radios, see https://github.com/jgromes/RadioLib/tree/master/examples/AFSK/AFSK_External_Radio

**Comment by ManoDaSilva** [comment]
Hi Sorry for the delay! That's really nice! Now we can use the full library with a nice 25W radio :) 

Regarding 2-FSK, isn't there a way to toggle a pin instead of changing registers, and redirect that to whatever external transmitter we're using? Would that imply some timing issues?

**Reply by jgromes** [reply]
How exactly would you determine the frequency at which to toggle the pin?

An example: POCSAG specifies frequency deviation of +- 4.5 kHz between the 1s and 0s. What should be the freequencies the pin toggles at?

**Comment by ManoDaSilva** [comment]
The FSK deviation (when driving an FM radio via a packet/unfiltered input) is adjusted by varying the amplitude of the input signal, not its frequency... The toggling "speed" will dictate the baud rate :)

**Reply by jgromes** [reply]
Varying the amplitude is somewhat challenging on a digital output pin. That would effectively limit this to Arudino boards with DAC. Effectively, the only thing that can be changed here is the frequency, hence why I think this will remain limited to AFSK. If your radio can take audio input, that is.

**Reply by ManoDaSilva** [reply]
I was seeing something like the digital output being fed to either a simple amp or a resistor divider (depending on whether it needs amplification or attenuation) to bring it to the appropriate level for the desired deviation, it does require some additional components, but I believe it's still quite simple to set up. Once I've got some free time, I'll give it a go to see how it does.

**Comment by ManoDaSilva** [comment]
Okay, 7 months later, I came back to this issue as I'm still very interested in using this lib for directly driving my radios for 2-FSK :) 

Granted I still need to adjust the level after the output pin via a simple potentiometer.

I tried to implement it using the ExternalRadio Protocol, and modding Pager.cpp to handle driving a pin directly instead of changing the registers. Most of the tweaks were about making sure it made no calls to the Physical Layer for frequency settings, deviation, etc. I also added the pin toggling function where the registry access was.

My implementation is very hacky, but I managed to get a square signal that looks exactly like the signal I usually feed into my radio using Unipager, when transmitting a simple "Tone" message. 
Weirdly enough, when transmitting a BCD or ASCII message, my ATmeg328 resets. I wonder if it has anything to do with memory running out, I'll try it on an ESP32 to confirm my doubts.

**Reply by jgromes** [reply]
> My implementation is very hacky, but I managed to get a square signal that looks exactly like the signal I usually feed into my radio using Unipager, when transmitting a simple "Tone" message.

I would be interested to see that, is it just logic 0/1 for low/high frequency, or is it audio-modulated?

> Weirdly enough, when transmitting a BCD or ASCII message, my ATmeg328 resets

It's possible, ATmega328 has very limited program and data memory. So much so that I'm thinking of dropping the support altogether. What is the program/data usage when you are uploading?

**Reply by ManoDaSilva** [reply]
> I would be interested to see that, is it just logic 0/1 for low/high frequency, or is it audio-modulated?

Logic 0/1, I'm just using mod->hal->digitalWrite

> It's possible, ATmega328 has very limited program and data memory. So much so that I'm thinking of dropping the support altogether. What is the program/data usage when you are uploading?

`Sketch uses 13608 bytes (44%) of program storage space. Maximum is 30720 bytes.
Global variables use 726 bytes (35%) of dynamic memory, leaving 1322 bytes for local variables. Maximum is 2048 bytes.`
I also suspect the ATmega is showing its limits. I'll try on an ESP32, see what it does.

**Reply by jgromes** [reply]
Seems like that should be plenty to run the program, so I'm guessing there's sometihing wrong with memory mangement. Can you post your Arduino sketch?

**Reply by ManoDaSilva** [reply]
[RadioLib_DummyTx.txt](https://github.com/jgromes/RadioLib/files/11957989/RadioLib_DummyTx.txt)
(Had to rename my Arduino Sketch)

I also modified Pager.cpp and .h to implement handling I/Os. Tried to implement the External Radio management by checking the _RADIOLIB_EXTERNAL_RADIO_H define, but it doesn't seem to be handled by the preprocessor when declared in ExternalRadio.h, so I had to force it in Pager.cpp for now

[Pager.h.txt](https://github.com/jgromes/RadioLib/files/11958013/Pager.h.txt)
[Pager.cpp.txt](https://github.com/jgromes/RadioLib/files/11958014/Pager.cpp.txt)

**Reply by jgromes** [reply]
I don't see anything immediately obvious that could be causing a crash, so I'll investigate a bit more.

Regarding the external radio, a better approach would be to implement the `transmitDirect` method in the `ExternalRadio` class. I'll do that and let you know.

**Reply by jgromes** [reply]
@ManoDaSilva I found and fixed a memory leak in the BCH encoder, everything is now stable on ATmega328P. In addition, I have also implemented `transmitDirect`, so now you can do the following:

```c++
#include <RadioLib.h>

// output on pin 4
ExternalRadio radio(4);
PagerClient pager(&radio);

void setup() {
  pager.begin(434.0, 1200);
}
```

I examined the output with an oscilloscope and can clearly see the preamble followed by the frame synchronization word. One thing I noticed is that the waveform is inverted compared to the binary data, so if it doesn't work the first time I suggest checking the inversion setting.

**Reply by ManoDaSilva** [reply]
Good copy! I will try it this week :) 

Thank you for the hard work!
